probe/endpoint/procspy/background_reader_linux.go

```diff
@@ -0,0 +1,144 @@
+	defer br.mtx.Unlock()
+
+	_, err := io.Copy(buf, br.latestBuf)
+
+	return br.latestSockets, err
+}
+
+type walkResult struct {
+	buf     *bytes.Buffer
+	sockets map[uint64]*Proc
+}
+
+func performWalk(w pidWalker, c chan<- walkResult) {
+	var (
+		err    error
+		result = walkResult{
+			buf: bytes.NewBuffer(make([]byte, 0, 5000)),
+		}
+	)
+
+	result.sockets, err = w.walk(result.buf)
+	if err != nil {
+		log.Errorf("background /proc reader: error walking /proc: %s", err)
+		result.buf.Reset()
+		result.sockets = nil
+	}
+	c <- result
+}
+
+func (br *backgroundReader) loop(walker process.Walker) {
+	var (
+		begin           time.Time                      // when we started the last performWalk
+		tickc           = time.After(time.Millisecond) // fire immediately
+		walkc           chan walkResult                // initially nil, i.e. off
+		rateLimitPeriod = initialRateLimitPeriod
+		restInterval    time.Duration
+		ticker          = time.NewTicker(rateLimitPeriod)
+		pWalker         = newPidWalker(walker, ticker.C, fdBlockSize)
+	)
+
+	for {
+		select {
+		case <-tickc:
+			tickc = nil                      // turn off until the next loop
+			walkc = make(chan walkResult, 1) // turn on (need buffered so we don't leak performWalk)
+			begin = time.Now()               // reset counter
+			go performWalk(pWalker, walkc)   // do work
+
+		case result := <-walkc:
+			// Expose results
+			br.mtx.Lock()
+			br.latestBuf = result.buf
+			br.latestSockets = result.sockets
+			br.mtx.Unlock()
+
+			// Schedule next walk and adjust its rate limit
+			walkTime := time.Since(begin)
+			rateLimitPeriod, restInterval = scheduleNextWalk(rateLimitPeriod, walkTime)
+			ticker.Stop()
+			ticker = time.NewTicker(rateLimitPeriod)
+			pWalker.tickc = ticker.C
+
+			walkc = nil                      // turn off until the next loop
+			tickc = time.After(restInterval) // turn on
+
+		case <-br.stopc:
+			pWalker.stop()
+			ticker.Stop()
+			return // abort
+		}
+	}
+}
+
+// Adjust rate limit for next walk and calculate when it should be started
+func scheduleNextWalk(rateLimitPeriod time.Duration, took time.Duration) (newRateLimitPeriod time.Duration, restInterval time.Duration) {
+	log.Debugf("background /proc reader: full pass took %s", took)
+	if float64(took)/float64(targetWalkTime) > 1.5 {
+		log.Warnf(
+			"background /proc reader: full pass took %s: 50%% more than expected (%s)",
+			took,
+			targetWalkTime,
+		)
+	}
+
+	// Adjust rate limit to more-accurately meet the target walk time in next iteration
+	newRateLimitPeriod = time.Duration(float64(targetWalkTime) / float64(took) * float64(rateLimitPeriod))
+	if newRateLimitPeriod > maxRateLimitPeriod {
+		newRateLimitPeriod = maxRateLimitPeriod
+	} else if newRateLimitPeriod < minRateLimitPeriod {
+		newRateLimitPeriod = minRateLimitPeriod
+	}
+	log.Debugf("background /proc reader: new rate limit period %s", newRateLimitPeriod)
+
+	// when to start next walk
+	restInterval = targetWalkTime - took
+
+	return
```

Comment:
This is a so-called naked return, which is best avoided.

> Naked returns are okay if the function is a handful of lines. Once it's a medium-sized function, be explicit with your return values.
—https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters